File: src/internalDebug/groovy/cash/andrew/mntrailconditions/ui/debug/DebugView.java
package cash.andrew.mntrailconditions.ui.debug;

import android.animation.ValueAnimator;
import android.app.Application;
import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.support.v7.app.AlertDialog;
import android.util.AttributeSet;
import android.util.DisplayMetrics;
import android.view.ContextThemeWrapper;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.EditText;
import android.widget.FrameLayout;
import android.widget.Spinner;
import android.widget.Switch;
import android.widget.TextView;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import cash.andrew.mntrailconditions.util.IntentManager;
import com.f2prateek.rx.preferences.Preference;
import com.jakewharton.processphoenix.ProcessPhoenix;
import com.jakewharton.rxbinding.widget.RxAdapterView;
import cash.andrew.mntrailconditions.BuildConfig;
import cash.andrew.mntrailconditions.R;
import cash.andrew.mntrailconditions.data.AnimationSpeed;
import cash.andrew.mntrailconditions.data.ApiEndpoint;
import cash.andrew.mntrailconditions.data.ApiEndpoints;
import cash.andrew.mntrailconditions.data.CaptureIntents;
import cash.andrew.mntrailconditions.data.Injector;
import cash.andrew.mntrailconditions.data.LumberYard;
import cash.andrew.mntrailconditions.data.NetworkDelay;
import cash.andrew.mntrailconditions.data.NetworkFailurePercent;
import cash.andrew.mntrailconditions.data.NetworkVariancePercent;
import cash.andrew.mntrailconditions.data.PicassoDebugging;
import cash.andrew.mntrailconditions.data.PixelGridEnabled;
import cash.andrew.mntrailconditions.data.PixelRatioEnabled;
import cash.andrew.mntrailconditions.data.ScalpelEnabled;
import cash.andrew.mntrailconditions.data.ScalpelWireframeEnabled;
import cash.andrew.mntrailconditions.data.prefs.InetSocketAddressPreferenceAdapter;
import cash.andrew.mntrailconditions.ui.logs.LogsDialog;
import cash.andrew.mntrailconditions.ui.misc.EnumAdapter;
import cash.andrew.mntrailconditions.util.Keyboards;
import cash.andrew.mntrailconditions.util.Strings;
import com.squareup.leakcanary.internal.DisplayLeakActivity;
import com.squareup.picasso.Picasso;
import com.squareup.picasso.StatsSnapshot;
import java.lang.reflect.Method;
import java.net.InetSocketAddress;
import java.util.Locale;
import javax.inject.Inject;
import javax.inject.Named;
import okhttp3.Cache;
import okhttp3.OkHttpClient;
import org.threeten.bp.Instant;
import org.threeten.bp.ZoneId;
import org.threeten.bp.format.DateTimeFormatter;
import org.threeten.bp.temporal.TemporalAccessor;
import retrofit2.mock.NetworkBehavior;
import timber.log.Timber;

import static butterknife.ButterKnife.findById;

public final class DebugView extends FrameLayout {
  private static final DateTimeFormatter DATE_DISPLAY_FORMAT =
      DateTimeFormatter.ofPattern("yyyy-MM-dd hh:mm a", Locale.US).withZone(ZoneId.systemDefault());

  @BindView(R.id.debug_network_endpoint) Spinner endpointView;
  @BindView(R.id.debug_network_endpoint_edit) View endpointEditView;
  @BindView(R.id.debug_network_proxy) Spinner networkProxyView;

  @BindView(R.id.debug_ui_animation_speed) Spinner uiAnimationSpeedView;
  @BindView(R.id.debug_ui_pixel_grid) Switch uiPixelGridView;
  @BindView(R.id.debug_ui_pixel_ratio) Switch uiPixelRatioView;
  @BindView(R.id.debug_ui_scalpel) Switch uiScalpelView;
  @BindView(R.id.debug_ui_scalpel_wireframe) Switch uiScalpelWireframeView;

  @BindView(R.id.debug_build_name) TextView buildNameView;
  @BindView(R.id.debug_build_code) TextView buildCodeView;
  @BindView(R.id.debug_build_sha) TextView buildShaView;
  @BindView(R.id.debug_build_date) TextView buildDateView;

  @BindView(R.id.debug_device_make) TextView deviceMakeView;
  @BindView(R.id.debug_device_model) TextView deviceModelView;
  @BindView(R.id.debug_device_resolution) TextView deviceResolutionView;
  @BindView(R.id.debug_device_density) TextView deviceDensityView;
  @BindView(R.id.debug_device_release) TextView deviceReleaseView;
  @BindView(R.id.debug_device_api) TextView deviceApiView;

  @BindView(R.id.debug_picasso_indicators) Switch picassoIndicatorView;
  @BindView(R.id.debug_picasso_cache_size) TextView picassoCacheSizeView;
  @BindView(R.id.debug_picasso_cache_hit) TextView picassoCacheHitView;
  @BindView(R.id.debug_picasso_cache_miss) TextView picassoCacheMissView;
  @BindView(R.id.debug_picasso_decoded) TextView picassoDecodedView;
  @BindView(R.id.debug_picasso_decoded_total) TextView picassoDecodedTotalView;
  @BindView(R.id.debug_picasso_decoded_avg) TextView picassoDecodedAvgView;
  @BindView(R.id.debug_picasso_transformed) TextView picassoTransformedView;
  @BindView(R.id.debug_picasso_transformed_total) TextView picassoTransformedTotalView;
  @BindView(R.id.debug_picasso_transformed_avg) TextView picassoTransformedAvgView;

  @BindView(R.id.debug_okhttp_cache_max_size) TextView okHttpCacheMaxSizeView;
  @BindView(R.id.debug_okhttp_cache_write_error) TextView okHttpCacheWriteErrorView;
  @BindView(R.id.debug_okhttp_cache_request_count) TextView okHttpCacheRequestCountView;
  @BindView(R.id.debug_okhttp_cache_network_count) TextView okHttpCacheNetworkCountView;
  @BindView(R.id.debug_okhttp_cache_hit_count) TextView okHttpCacheHitCountView;

  @Inject OkHttpClient client;
  @Inject @Named("Api") OkHttpClient apiClient;
  @Inject Picasso picasso;
  @Inject LumberYard lumberYard;
  @Inject @ApiEndpoint Preference<String> networkEndpoint;
  @Inject Preference<InetSocketAddress> networkProxyAddress;
  @Inject @CaptureIntents Preference<Boolean> captureIntents;
  @Inject @AnimationSpeed Preference<Integer> animationSpeed;
  @Inject @PicassoDebugging Preference<Boolean> picassoDebugging;
  @Inject @PixelGridEnabled Preference<Boolean> pixelGridEnabled;
  @Inject @PixelRatioEnabled Preference<Boolean> pixelRatioEnabled;
  @Inject @ScalpelEnabled Preference<Boolean> scalpelEnabled;
  @Inject @ScalpelWireframeEnabled Preference<Boolean> scalpelWireframeEnabled;
  @Inject NetworkBehavior behavior;
  @Inject @NetworkDelay Preference<Long> networkDelay;
  @Inject @NetworkFailurePercent Preference<Integer> networkFailurePercent;
  @Inject @NetworkVariancePercent Preference<Integer> networkVariancePercent;
  @Inject Application app;
  @Inject IntentManager intentManager;

  public DebugView(Context context) {
    this(context, null);
  }

  public DebugView(Context context, AttributeSet attrs) {
    super(context, attrs);
    Injector.obtain(context).inject(this);

    // Inflate all of the controls and inject them.
    LayoutInflater.from(context).inflate(R.layout.debug_view_content, this);
    ButterKnife.bind(this);

    setupNetworkSection();
    setupUserInterfaceSection();
    setupBuildSection();
    setupDeviceSection();
    setupPicassoSection();
    setupOkHttpCacheSection();
  }

  public void onDrawerOpened() {
    refreshPicassoStats();
    refreshOkHttpCacheStats();
  }

  private void setupNetworkSection() {
    final ApiEndpoints currentEndpoint = ApiEndpoints.from(networkEndpoint.get());
    final EnumAdapter<ApiEndpoints> endpointAdapter =
        new EnumAdapter<>(getContext(), ApiEndpoints.class);
    endpointView.setAdapter(endpointAdapter);
    endpointView.setSelection(currentEndpoint.ordinal());

    RxAdapterView.itemSelections(endpointView)
        .map(endpointAdapter::getItem)
        .filter(item -> item != currentEndpoint)
        .subscribe(selected -> {
          if (selected == ApiEndpoints.CUSTOM) {
            Timber.d("Custom network endpoint selected. Prompting for URL.");
            showCustomEndpointDialog(currentEndpoint.ordinal(), "http://");
          } else {
            setEndpointAndRelaunch(selected.url);
          }
        });

    int currentProxyPosition = networkProxyAddress.isSet() ? ProxyAdapter.PROXY : ProxyAdapter.NONE;
    final ProxyAdapter proxyAdapter = new ProxyAdapter(getContext(), networkProxyAddress);
    networkProxyView.setAdapter(proxyAdapter);
    networkProxyView.setSelection(currentProxyPosition);

    RxAdapterView.itemSelections(networkProxyView)
        .filter(position -> !networkProxyAddress.isSet() || position != ProxyAdapter.PROXY)
        .subscribe(position -> {
          if (position == ProxyAdapter.NONE) {
            // Only clear the proxy and restart if one was previously set.
            if (currentProxyPosition != ProxyAdapter.NONE) {
              Timber.d("Clearing network proxy");
              // TODO: Keep the custom proxy around so you can easily switch back and forth.
              networkProxyAddress.delete();
              // Force a restart to re-initialize the app without a proxy.
              ProcessPhoenix.triggerRebirth(getContext());
            }
          } else if (networkProxyAddress.isSet() && position == ProxyAdapter.PROXY) {
            Timber.d("Ignoring re-selection of network proxy %s", networkProxyAddress.get());
          } else {
            Timber.d("New network proxy selected. Prompting for host.");
            showNewNetworkProxyDialog(proxyAdapter);
          }
        });

    // Only show the endpoint editor when a custom endpoint is in use.
    endpointEditView.setVisibility(currentEndpoint == ApiEndpoints.CUSTOM ? VISIBLE : GONE);
  }

  @OnClick(R.id.debug_network_endpoint_edit) void onEditEndpointClicked() {
    Timber.d("Prompting to edit custom endpoint URL.");
    // Pass in the currently selected position since we are merely editing.
    showCustomEndpointDialog(endpointView.getSelectedItemPosition(), networkEndpoint.get());
  }

  private void setupUserInterfaceSection() {
    final AnimationSpeedAdapter speedAdapter = new AnimationSpeedAdapter(getContext());
    uiAnimationSpeedView.setAdapter(speedAdapter);
    final int animationSpeedValue = animationSpeed.get();
    uiAnimationSpeedView.setSelection(
        AnimationSpeedAdapter.getPositionForValue(animationSpeedValue));

    RxAdapterView.itemSelections(uiAnimationSpeedView)
        .map(speedAdapter::getItem)
        .filter(item -> item != animationSpeed.get())
        .subscribe(selected -> {
          Timber.d("Setting animation speed to %sx", selected);
          animationSpeed.set(selected);
          applyAnimationSpeed(selected);
        });
    // Ensure the animation speed value is always applied across app restarts.
    post(() -> applyAnimationSpeed(animationSpeedValue));

    boolean gridEnabled = pixelGridEnabled.get();
    uiPixelGridView.setChecked(gridEnabled);
    uiPixelRatioView.setEnabled(gridEnabled);
    uiPixelGridView.setOnCheckedChangeListener((buttonView, isChecked) -> {
      Timber.d("Setting pixel grid overlay enabled to %b", isChecked);
      pixelGridEnabled.set(isChecked);
      uiPixelRatioView.setEnabled(isChecked);
    });

    uiPixelRatioView.setChecked(pixelRatioEnabled.get());
    uiPixelRatioView.setOnCheckedChangeListener((buttonView, isChecked) -> {
      Timber.d("Setting pixel scale overlay enabled to %b", isChecked);
      pixelRatioEnabled.set(isChecked);
    });

    uiScalpelView.setChecked(scalpelEnabled.get());
    uiScalpelWireframeView.setEnabled(scalpelEnabled.get());
    uiScalpelView.setOnCheckedChangeListener((buttonView, isChecked) -> {
      Timber.d("Setting scalpel interaction enabled to %b", isChecked);
      scalpelEnabled.set(isChecked);
      uiScalpelWireframeView.setEnabled(isChecked);
    });

    uiScalpelWireframeView.setChecked(scalpelWireframeEnabled.get());
    uiScalpelWireframeView.setOnCheckedChangeListener((buttonView, isChecked) -> {
      Timber.d("Setting scalpel wireframe enabled to %b", isChecked);
      scalpelWireframeEnabled.set(isChecked);
    });
  }

  @OnClick(R.id.debug_logs_show) void showLogs() {
    new LogsDialog(new ContextThemeWrapper(getContext(), R.style.Theme_MnTrailConditions), lumberYard, intentManager).show();
  }

  @OnClick(R.id.debug_leaks_show) void showLeaks() {
    Intent intent = new Intent(getContext(), DisplayLeakActivity.class);
    getContext().startActivity(intent);
  }

  private void setupBuildSection() {
    buildNameView.setText(BuildConfig.VERSION_NAME);
    buildCodeView.setText(String.valueOf(BuildConfig.VERSION_CODE));
    buildShaView.setText(BuildConfig.GIT_SHA);

    TemporalAccessor buildTime = Instant.ofEpochSecond(BuildConfig.GIT_TIMESTAMP);
    buildDateView.setText(DATE_DISPLAY_FORMAT.format(buildTime));
  }

  private void setupDeviceSection() {
    DisplayMetrics displayMetrics = getContext().getResources().getDisplayMetrics();
    String densityBucket = getDensityString(displayMetrics);
    deviceMakeView.setText(Strings.truncateAt(Build.MANUFACTURER, 20));
    deviceModelView.setText(Strings.truncateAt(Build.MODEL, 20));
    deviceResolutionView.setText(displayMetrics.heightPixels + "x" + displayMetrics.widthPixels);
    deviceDensityView.setText(displayMetrics.densityDpi + "dpi (" + densityBucket + ")");
    deviceReleaseView.setText(Build.VERSION.RELEASE);
    deviceApiView.setText(String.valueOf(Build.VERSION.SDK_INT));
  }

  private void setupPicassoSection() {
    boolean picassoDebuggingValue = picassoDebugging.get();
    picasso.setIndicatorsEnabled(picassoDebuggingValue);
    picassoIndicatorView.setChecked(picassoDebuggingValue);
    picassoIndicatorView.setOnCheckedChangeListener((button, isChecked) -> {
      Timber.d("Setting Picasso debugging to " + isChecked);
      picasso.setIndicatorsEnabled(isChecked);
      picassoDebugging.set(isChecked);
    });

    refreshPicassoStats();
  }

  private void refreshPicassoStats() {
    StatsSnapshot snapshot = picasso.getSnapshot();
    String size = getSizeString(snapshot.size);
    String total = getSizeString(snapshot.maxSize);
    int percentage = (int) ((1f * snapshot.size / snapshot.maxSize) * 100);
    picassoCacheSizeView.setText(size + " / " + total + " (" + percentage + "%)");
    picassoCacheHitView.setText(String.valueOf(snapshot.cacheHits));
    picassoCacheMissView.setText(String.valueOf(snapshot.cacheMisses));
    picassoDecodedView.setText(String.valueOf(snapshot.originalBitmapCount));
    picassoDecodedTotalView.setText(getSizeString(snapshot.totalOriginalBitmapSize));
    picassoDecodedAvgView.setText(getSizeString(snapshot.averageOriginalBitmapSize));
    picassoTransformedView.setText(String.valueOf(snapshot.transformedBitmapCount));
    picassoTransformedTotalView.setText(getSizeString(snapshot.totalTransformedBitmapSize));
    picassoTransformedAvgView.setText(getSizeString(snapshot.averageTransformedBitmapSize));
  }

  private void setupOkHttpCacheSection() {
    Cache cache = client.cache(); // Shares the cache with apiClient, so no need to check both.
    okHttpCacheMaxSizeView.setText(getSizeString(cache.maxSize()));

    refreshOkHttpCacheStats();
  }

  private void refreshOkHttpCacheStats() {
    Cache cache = client.cache(); // Shares the cache with apiClient, so no need to check both.
    int writeTotal = cache.writeSuccessCount() + cache.writeAbortCount();
    int percentage = (int) ((1f * cache.writeAbortCount() / writeTotal) * 100);
    okHttpCacheWriteErrorView.setText(
        cache.writeAbortCount() + " / " + writeTotal + " (" + percentage + "%)");
    okHttpCacheRequestCountView.setText(String.valueOf(cache.requestCount()));
    okHttpCacheNetworkCountView.setText(String.valueOf(cache.networkCount()));
    okHttpCacheHitCountView.setText(String.valueOf(cache.hitCount()));
  }

  private void applyAnimationSpeed(int multiplier) {
    try {
      Method method = ValueAnimator.class.getDeclaredMethod("setDurationScale", float.class);
      method.invoke(null, (float) multiplier);
    } catch (Exception e) {
      throw new RuntimeException("Unable to apply animation speed.", e);
    }
  }

  private static String getDensityString(DisplayMetrics displayMetrics) {
    switch (displayMetrics.densityDpi) {
      case DisplayMetrics.DENSITY_LOW:
        return "ldpi";
      case DisplayMetrics.DENSITY_MEDIUM:
        return "mdpi";
      case DisplayMetrics.DENSITY_HIGH:
        return "hdpi";
      case DisplayMetrics.DENSITY_XHIGH:
        return "xhdpi";
      case DisplayMetrics.DENSITY_XXHIGH:
        return "xxhdpi";
      case DisplayMetrics.DENSITY_XXXHIGH:
        return "xxxhdpi";
      case DisplayMetrics.DENSITY_TV:
        return "tvdpi";
      default:
        return String.valueOf(displayMetrics.densityDpi);
    }
  }

  private static String getSizeString(long bytes) {
    String[] units = new String[] { "B", "KB", "MB", "GB" };
    int unit = 0;
    while (bytes >= 1024) {
      bytes /= 1024;
      unit += 1;
    }
    return bytes + units[unit];
  }

  private void showNewNetworkProxyDialog(final ProxyAdapter proxyAdapter) {
    final int originalSelection = networkProxyAddress.isSet() ? ProxyAdapter.PROXY : ProxyAdapter.NONE;

    View view = LayoutInflater.from(app).inflate(R.layout.debug_drawer_network_proxy, null);
    final EditText hostView = findById(view, R.id.debug_drawer_network_proxy_host);

    if (networkProxyAddress.isSet()) {
      String host = networkProxyAddress.get().getHostName();
      hostView.setText(host); // Set the current host.
      hostView.setSelection(0, host.length()); // Pre-select it for editing.

      // Show the keyboard. Post this to the next frame when the dialog has been attached.
      hostView.post(() -> Keyboards.showKeyboard(hostView));
    }

    new AlertDialog.Builder(getContext()) //
        .setTitle("Set Network Proxy")
        .setView(view)
        .setNegativeButton("Cancel", (dialog, i) -> {
          networkProxyView.setSelection(originalSelection);
          dialog.cancel();
        })
        .setPositiveButton("Use", (dialog, i) -> {
          String in = hostView.getText().toString();
          InetSocketAddress address = InetSocketAddressPreferenceAdapter.parse(in);
          if (address != null) {
            networkProxyAddress.set(address);
            // Force a restart to re-initialize the app with the new proxy.
            ProcessPhoenix.triggerRebirth(getContext());
          } else {
            networkProxyView.setSelection(originalSelection);
          }
        })
        .setOnCancelListener(dialogInterface -> networkProxyView.setSelection(originalSelection))
        .show();
  }

  private void showCustomEndpointDialog(final int originalSelection, String defaultUrl) {
    View view = LayoutInflater.from(app).inflate(R.layout.debug_drawer_network_endpoint, null);
    final EditText url = findById(view, R.id.debug_drawer_network_endpoint_url);
    url.setText(defaultUrl);
    url.setSelection(url.length());

    new AlertDialog.Builder(getContext()) //
        .setTitle("Set Network Endpoint")
        .setView(view)
        .setNegativeButton("Cancel", (dialog, i) -> {
          endpointView.setSelection(originalSelection);
          dialog.cancel();
        })
        .setPositiveButton("Use", (dialog, i) -> {
            String theUrl = url.getText().toString();
            if (!Strings.isBlank(theUrl)) {
              setEndpointAndRelaunch(theUrl);
            } else {
              endpointView.setSelection(originalSelection);
            }
        })
        .setOnCancelListener((dialogInterface) -> {
            endpointView.setSelection(originalSelection);
        })
        .show();
  }

  private void setEndpointAndRelaunch(String endpoint) {
    Timber.d("Setting network endpoint to %s", endpoint);
    networkEndpoint.set(endpoint);

    ProcessPhoenix.triggerRebirth(getContext());
  }
}

File: src/debug/java/cash/andrew/mntrailconditions/ui/debug/DebugView.kt
package cash.andrew.mntrailconditions.ui.debug

import android.annotation.SuppressLint
import android.app.Application
import android.content.Context
import android.os.Build
import android.util.AttributeSet
import android.util.DisplayMetrics
import android.view.LayoutInflater
import android.view.View
import android.widget.*
import androidx.appcompat.app.AlertDialog
import cash.andrew.mntrailconditions.BuildConfig
import cash.andrew.mntrailconditions.MnTrailConditionsApp
import cash.andrew.mntrailconditions.R
import cash.andrew.mntrailconditions.data.ApiEndpoint
import cash.andrew.mntrailconditions.data.ApiEndpoints
import cash.andrew.mntrailconditions.data.preference.Preference
import cash.andrew.mntrailconditions.databinding.DebugDrawerNetworkEndpointBinding
import cash.andrew.mntrailconditions.databinding.DebugViewContentBinding
import cash.andrew.mntrailconditions.ui.misc.EnumAdapter
import cash.andrew.mntrailconditions.util.IntentManager
import com.jakewharton.processphoenix.ProcessPhoenix
import com.readystatesoftware.chuck.Chuck
import okhttp3.OkHttpClient
import timber.log.Timber
import javax.inject.Inject

class DebugView
@JvmOverloads constructor(
        context: Context,
        attrs: AttributeSet? = null
) : GridLayout(context, attrs) {

    private val endpointView get() = binding.debugNetworkEndpoint
    private val endpointEditView get() = binding.debugNetworkEndpointEdit
    private val buildNameView get() = binding.debugBuildName
    private val buildCodeView get() = binding.debugBuildCode
    private val deviceMakeView get() = binding.debugDeviceMake
    private val deviceModelView get() = binding.debugDeviceModel
    private val deviceResolutionView get() = binding.debugDeviceResolution
    private val deviceDensityView get() = binding.debugDeviceDensity
    private val deviceReleaseView get() = binding.debugDeviceRelease
    private val deviceApiView get() = binding.debugDeviceApi
    private val okHttpCacheMaxSizeView get() = binding.debugOkhttpCacheMaxSize
    private val okHttpCacheWriteErrorView get() = binding.debugOkhttpCacheWriteError
    private val okHttpCacheRequestCountView get() = binding.debugOkhttpCacheRequestCount
    private val okHttpCacheNetworkCountView get() = binding.debugOkhttpCacheNetworkCount
    private val okHttpCacheHitCountView get() = binding.debugOkhttpCacheHitCount

    @Inject lateinit var client: OkHttpClient
    @Inject @field:ApiEndpoint lateinit var networkEndpoint: Preference<String>
    @Inject lateinit var app: Application
    @Inject lateinit var intentManager: IntentManager

    private val binding: DebugViewContentBinding

    init {
        (context.applicationContext as MnTrailConditionsApp).component.inject(this)
        View.inflate(context, R.layout.debug_view_content, this)

        // Inflate all of the controls and inject them.
        binding = DebugViewContentBinding.bind(this)

        setupNetworkSection()
        setupBuildSection()
        setupDeviceSection()
        setupOkHttpCacheSection()

        endpointEditView.setOnClickListener {
            Timber.d("Prompting to edit custom endpoint URL.")
            // Pass in the currently selected position since we are merely editing.
            showCustomEndpointDialog(endpointView.selectedItemPosition, networkEndpoint.get())
        }

        binding.debugNetworkLogsShow.setOnClickListener {
            val intent = Chuck.getLaunchIntent(app)
            app.startActivity(intent)
        }
    }

    fun onDrawerOpened() {
        refreshOkHttpCacheStats()
    }

    @SuppressLint("CheckResult")
    private fun setupNetworkSection() {
        val currentEndpoint = ApiEndpoints.from(networkEndpoint.get())
        val endpointAdapter = EnumAdapter(context, ApiEndpoints::class.java)
        endpointView.adapter = endpointAdapter
        endpointView.setSelection(currentEndpoint.ordinal)

        endpointView.onItemSelectedListener = object: AdapterView.OnItemSelectedListener{
            override fun onNothingSelected(parent: AdapterView<*>?) {
            }

            override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
                val selected = endpointAdapter.getItem(position)
                if (selected == currentEndpoint) return

                if (selected === ApiEndpoints.CUSTOM) {
                    Timber.d("Custom network endpoint selected. Prompting for URL.")
                    showCustomEndpointDialog(currentEndpoint.ordinal, "http://")
                } else {
                    setEndpointAndRelaunch(selected!!.url)
                }
            }
        }

        // Only show the endpoint editor when a custom endpoint is in use.
        endpointEditView.visibility = if (currentEndpoint === ApiEndpoints.CUSTOM) View.VISIBLE else View.GONE
    }

    private fun setupBuildSection() {
        buildNameView.text = BuildConfig.VERSION_NAME
        buildCodeView.text = BuildConfig.VERSION_CODE.toString()
    }

    @SuppressLint("SetTextI18n")
    private fun setupDeviceSection() {
        val displayMetrics = context.resources.displayMetrics
        val densityBucket = getDensityString(displayMetrics)
        deviceMakeView.text = Build.MANUFACTURER.take(20)
        deviceModelView.text = Build.MODEL.take(20)
        deviceResolutionView.text = "${displayMetrics.heightPixels}x${displayMetrics.widthPixels}"
        deviceDensityView.text = "${displayMetrics.densityDpi}dpi ($densityBucket)"
        deviceReleaseView.text = Build.VERSION.RELEASE
        deviceApiView.text = Build.VERSION.SDK_INT.toString()
    }

    private fun setupOkHttpCacheSection() {
        val cache = client.cache!! // Shares the cache with apiClient, so no need to check both.
        okHttpCacheMaxSizeView.text = getSizeString(cache.maxSize())

        refreshOkHttpCacheStats()
    }

    @SuppressLint("SetTextI18n")
    private fun refreshOkHttpCacheStats() {
        val cache = client.cache!! // Shares the cache with apiClient, so no need to check both.
        val writeTotal = cache.writeSuccessCount() + cache.writeAbortCount()
        val percentage = (1f * cache.writeAbortCount() / writeTotal * 100).toInt()
        okHttpCacheWriteErrorView.text = "${cache.writeAbortCount()}/$writeTotal($percentage%)"
        okHttpCacheRequestCountView.text = cache.requestCount().toString()
        okHttpCacheNetworkCountView.text = cache.networkCount().toString()
        okHttpCacheHitCountView.text = cache.hitCount().toString()
    }

    @SuppressLint("InflateParams")
    private fun showCustomEndpointDialog(originalSelection: Int, defaultUrl: String) {
        val view = LayoutInflater.from(app).inflate(R.layout.debug_drawer_network_endpoint, null)
        val viewBinding = DebugDrawerNetworkEndpointBinding.bind(view)
        val url: EditText = viewBinding.debugDrawerNetworkEndpointUrl
        url.setText(defaultUrl)
        url.setSelection(url.length())

        AlertDialog.Builder(context) //
                .setTitle("Set Network Endpoint")
                .setView(view)
                .setNegativeButton("Cancel") { dialog, _ ->
                    endpointView.setSelection(originalSelection)
                    dialog.cancel()
                }
                .setPositiveButton("Use") { _, _ ->
                    val theUrl = url.text.toString()
                    if (theUrl.isNotBlank()) {
                        setEndpointAndRelaunch(theUrl)
                    } else {
                        endpointView.setSelection(originalSelection)
                    }
                }
                .setOnCancelListener { _ -> endpointView.setSelection(originalSelection) }
                .show()
    }

    @SuppressLint("CheckResult")
    private fun setEndpointAndRelaunch(endpoint: String?) {
        Timber.d("Setting network endpoint to %s", endpoint)
        networkEndpoint.set(endpoint!!)

        postDelayed({
            ProcessPhoenix.triggerRebirth(context)
        }, 1000)
    }

    private fun getDensityString(displayMetrics: DisplayMetrics): String {
        return when (displayMetrics.densityDpi) {
            DisplayMetrics.DENSITY_LOW -> "ldpi"
            DisplayMetrics.DENSITY_MEDIUM -> "mdpi"
            DisplayMetrics.DENSITY_HIGH -> "hdpi"
            DisplayMetrics.DENSITY_XHIGH -> "xhdpi"
            DisplayMetrics.DENSITY_XXHIGH -> "xxhdpi"
            DisplayMetrics.DENSITY_XXXHIGH -> "xxxhdpi"
            DisplayMetrics.DENSITY_TV -> "tvdpi"
            else -> displayMetrics.densityDpi.toString()
        }
    }

    private fun getSizeString(bytes: Long): String {
        var betterBytes = bytes
        val units = arrayOf("B", "KB", "MB", "GB")
        var unit = 0
        while (betterBytes >= 1024) {
            betterBytes /= 1024
            unit += 1
        }
        return betterBytes.toString() + units[unit]
    }
}
